Instructions: 

Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells)
It is suggested but not mandatory that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...

Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"

Save your python jupyter notebook in your own account and "repo" on github.com and submit a link to that notebook though Quercus for assignment marking

In [1]:
# In this chosen dataset, confirm that the dataset has missing values. 
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [2]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

I believe what this means is, for "id", there is 1 missing value. For "song", there are 11 missing values. The rest of the rows have no missing values. The other columns are complete. 

Pre-lecture Question #2

In [4]:
# In this chosen dataset, use what you've learned from ChatGPT to describe your dataset
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.columns # inspect all the columns and see all the column names in the data set

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

In [6]:
df.shape

(391, 11)

In [7]:
df.head

<bound method NDFrame.head of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye a

An "Observation" represents a unique, individual villager in the dataset, and each row provides information about a specific villager. 
A "Variable" is a column within a dataset, which represents various attributes/features about the villagers in Animal Crossing. 

Prelecture Question #3

In [8]:
df.shape # Provides number of rows, number of columns--which is a tuple

(391, 11)

In [9]:
df.info() # Provides basic information about the dataset 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


In [10]:
df.head() # Shows the first 5 rows of the dataset 

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [11]:
df.describe() # Shows summary of statistics of numerical variables. 

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [12]:
df.isna().sum() # Check for missing values in each column, already done but helpful in generating summary

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

Prelecture Question #4

In [13]:
url =  "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 

In [14]:
df.shape

(391, 11)

In [15]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [17]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.shape

(891, 15)

In [18]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Explanation of the discrepancy between the two descriptions of datasets: 

Because there are numerical columns with missing values, the count provided by df.describe() will inherently be less than the total number of rows in df.shape. Further, non-numerical columns won't appear in the df.describe() summary unless explicitly included. 
df.describe(), unless I include include='all', will only provide a summary of numerical columns with values that are not missing. 
This is why the two can give two different answers. 

Prelecture Question #5:

The difference between an "attribute", like df.shape which lacks (), and a "method", like df.describe() which includes (), is:
An attribute just stores values and doesn't compute anything, hence the lack of paranetheses. 
A method is a function that belong to something, and they do computation or actions. Thus, they must be called with paraentheses for proper execution. 

Now, we move onto post-lecture homework. 
Post-lecture question #6:

Count: The number of non-missing values for each column. 

Mean: The avg value of the data in the column. 

Standard Deviation: How much the data deviates from the mean; a dispersion. 

Minimum: Smallest value in the column. 

25%: The value below where 25% of the data falls (1st quartile).

50%: The median value (2nd quartile). 

75%: The value below where 75% of the data falls (3rd quartile). 

Maximum: The largest value in the column.

Is it clear what df.describe() is describing if there's missing values in the column it's analyzing? The answer is Yes: missing values are ignored when calculating statistics. Further, count tells you how many non-missing values were used in calculations, and you can always check the number of missing values using df.isna().sum(). 


Post-Lecture Question #7: 

1. Provide an example of a "use case" in which using df.dropna() might be preferred over using del df['col']. 

If you want to delete a column, irrespective of missing/non-missing values, then del df['col'] is preferred. If you want to delete columns or columns with missing values, then you'd use df.dropna(). 

An example is, in the Titanic dataset, a user might want to delete the columns/rows where there's missing data for the age. However, deleting the entire column of age would probably serve as a counterproductive measure since that is probably an important statistic for analyzing the profile of the victims who perished. 

2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna(). 

An opposite use case would be, if the user decides that the sex of the animal crossing characters is irrelevant to the final analysis, then they'd delete it entirely, irrespective of whether or not that column has missing values, non-missing values, or a combination of both. 

3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important. 

Because you'd probably want to remove junk columns that are not relevant first, reducing the complexity of the dataset. Also, you might remove important rows valuable for analysis if you apply df.dropna() to columns that are unnecessary. Using both together ensures you remove unnecessary data and have a handle on missing data, creating a complete dataset. 

4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna(), and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset. 


In [26]:
del df['sibsp']

In [27]:
df.dropna()

survived  pclass     sex   age     fare embarked  class    who  \
1           1       1  female  38.0  71.2833        C  First  woman   
3           1       1  female  35.0  53.1000        S  First  woman   
6           0       1    male  54.0  51.8625        S  First    man   
10          1       3  female   4.0  16.7000        S  Third  child   
11          1       1  female  58.0  26.5500        S  First  woman   
..        ...     ...     ...   ...      ...      ...    ...    ...   
871         1       1  female  47.0  52.5542        S  First  woman   
872         0       1    male  33.0   5.0000        S  First    man   
879         1       1  female  56.0  83.1583        C  First  woman   
887         1       1  female  19.0  30.0000        S  First  woman   
889         1       1    male  26.0  30.0000        C  First    man   

     adult_male deck alive  alone  
1         False    C   yes  False  
3         False    C   yes  False  
6          True    E    no   True  
10        False    G   yes  False  
11        False    C   yes   True  
..          ...  ...   ...    ...  
871       False    D   yes  False  
872        True    B    no   True  
879       False    C   yes  False  
887       False    B   yes   True  
889        True    C   yes   True  

[182 rows x 12 columns]

The goal that I had with the Titanic dataset was to wittle it down to important information for my specific goal, which was to have data that is relevant, or could be relevant, for analyzing the relationship of who died with social classifications such as class and genders. 
I did not need sibling pairs so I got rid of that first. Then, I got rid of all the missing values to give me a complete dataset. 

However, I classified every other column as important for a variety of questions I had in my mind, some of which include: 
What is the relationship between survival rate and class? Survival rate and gender? Survival rates of parent/children groups?

These questions would be part of a larger study (if I were to do one) about what the rates of survival could allude to (since correlation does not equal causation) about the morality, or social values, of the early 1900s. 


Post-Lecture Question #8: 

1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" dataset other than what the ChatBot automatically provides for you. 

This command gives detailed statistical summaries of column 2 with respect to each unique value of column 1. This allows the user to understand how the values of column 2 vary across various groups defined by column 1. 

An example is ('sex') and ('survived'):


In [28]:
df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

In this example, the key takeaway is the average number of females that survived the titanic was roughly 74.2%, whereas the average number of males that survived was roughly 18.9%. 

2. Assuming you've not yet removed missing values in the manner of question 7 above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()? 

The difference between df.describe() and df.groupby("col1")["col2"].describe() is, df.describe() provides a summary of all numerical/statistical data across all columns. df.groupby("col1")["col2"].describe() provides a numerical/statistical summary of only two columns of your choice. 

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to use (a) a ChatBot session to fix the errors, or (b) use google to search for and fix errors. 

a.) Forgetting to include import pandas as pd in code

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

NameError: name 'pd' is not defined

In [3]:
import pandas as pd # ChatGPT gave this as a solution to put this at the top of my code immediately
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

In this case, I had to look at two different sites on Google to find someone/an example that had the same error as me. For ChatGPT, it provided the solution immediately and clearly and I did not have to sift through anything.(a): ChatGPT solved this error better than Google 

b.) Mistype "titanic.csv" as "titanics.csv"

In [5]:
import pandas as pd # ChatGPT gave this as a solution to put this at the top of my code immediately
url = "titanics.csv"
df = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [8]:
import pandas as pd 
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

HTTPError: HTTP Error 404: Not Found

In replacing the entire URL with just "titanics.csv," I found both Google and ChatGPT helpful with the FileNotFound error. Google told me that there's no such file with said name, and to double-check the file name, which is pretty much what ChatGPT said. 
So if I ever made such an error, both Google and ChatGPT would lead me to double-check the file name. However, what gave ChatGPT an edge was the examples given. ChatGPT gave me examples of different file names (subdirectories, url, or absolute path). With this, I feel like I'd be able to trouble-shoot a bit better. (a), ChatGPT > Google. 

For the second error, where I just added an "s" to the titanics URL and got an HTTP error, ChatGPT won by a landslide. The AI gave me the correct URL and troubleshooted that I made a typo (based on the prior information I gave it beforehand). If I hadn't done so, and Chat had no prior history, it clearly defined what an HTTPError was and possible reasons for said error. Maybe my Google search wasn't specific enough for this, but the results would have confused me even more and used knowledge/jargon outside of my scope of knowledge. (a), ChatGPT > Google.

C.) Try to use a dataframe before it's been assigned into the variable. 

In [9]:
DF.groupby("sex")["age"].describe()

NameError: name 'DF' is not defined

Both Google and ChatGPT explicitly told me very quickly (not much sifting had to be done on Google for this) that my variable was not defined, and I had to double-check my function. However, once again, ChatGPT provided examples tailored to me and therefore was more relevant than a random batch of code on Stack overflow. (a), ChatGPT > Google 

D. Forget one of the parenthesis somewhere in the code. 

In [10]:
df.dropna(

SyntaxError: incomplete input (3039420378.py, line 1)

Both Google and ChatGPT were pretty quick to suggest I was either missing a colon, semi-colon, or parentheses. (a) = (b). 

F. Use a column name that's not in your data for the groupby and column section. 



In [12]:
df.groupby("Age")["sex"].describe()

KeyError: 'Age'

(a). ChatGPT>Google. ChatGPT suggested a typo error as one of its first troubleshooting suggestions, whereas I did not see a website in Google (not saying there isn't one, but I stopped searching after website 3) that suggested a typo instead of a complicated error. 

G. Forget to put a column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question. 


In [15]:
import pandas as pd 
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

In [19]:
df.groupby(parch)["sex"].describe()

NameError: name 'parch' is not defined

In [20]:
df.groupby(sex)["parch"].describe()

NameError: name 'sex' is not defined

I found ChatGPT to be very helpful, since the first suggestion was ensuring that I spelled everything correctly, and in code, spelling includes things like quotations. 
Google has a variety of different reasons why "NameError" occurs, and often the first suggestion is a basic misspelling. 
In this case, both were helpful since they'd direct me to first double-check the spelling of the word. 
(a) = (b). 

Post-Lecture Question #9: Yes. 

Summary of What I discussed with ChatGPT: https://chatgpt.com/share/66e3918b-6b90-8003-821a-a3c72c9a1783

In case for whatever reason this link is not working, here is a summary of what me and ChatGPT discussed: 

Certainly! Here’s a summary of the topics and issues we’ve discussed today:

1. Dataset Analysis:
Attributes vs. Methods: In pandas, df.shape is an attribute (doesn't use parentheses) that provides the dimensions of the DataFrame, while df.describe() is a method (uses parentheses) that provides summary statistics of the DataFrame.
Observations and Variables:
Observations: Each row in a dataset represents an observation.
Variables: Each column in a dataset represents a variable.
df.describe() Summary Statistics:
Count: Number of non-null entries in the column.
Mean: Average value of the column.
Std: Standard deviation, measuring the variability.
Min: Minimum value.
25%: 25th percentile value.
50%: Median or 50th percentile value.
75%: 75th percentile value.
Max: Maximum value.
Handling Missing Data:
df.dropna(): Removes rows with missing values.
del df['col']: Deletes the entire column from the DataFrame.
Use Case: Use del df['col'] when you want to remove an entire column that might be irrelevant, while df.dropna() is used to handle missing values row-wise.

2. Pandas Errors and Functions:
KeyError: Occurs when trying to access a column that does not exist or is misspelled.
Example Fix: Ensure correct column name by printing df.columns.
FileNotFoundError: Indicates the file path or name is incorrect.
Example Fix: Check and correct the file path or URL.
NameError: Happens when trying to use a variable or function that hasn’t been defined.
Example Fix: Ensure variables are defined before use and correctly spelled.
SyntaxError: Indicates an issue with code structure, like missing parentheses or incorrect indentation.
Example Fix: Review and correct the code structure and syntax.

3. Use of Methods and Attributes:
Attributes: Provide information about the DataFrame without needing parentheses, e.g., df.shape.
Methods: Perform operations on the DataFrame and require parentheses, e.g., df.describe().

4. Working with the Titanic Dataset:
Accessing Columns: Verify column names (e.g., df['age'] vs. df['Age']).
Grouping Data: df.groupby("col1")["col2"].describe() provides summary statistics of col2 grouped by col1.

5. Error Handling:
HTTPError: Indicates a problem with fetching data from a URL. Verify the URL and network connection.
Handling Missing Columns: Use df.get('column_name') to avoid errors when columns might be missing.
Summary
Understanding how to properly access and manipulate data in pandas is crucial for effective data analysis. Ensuring correct variable names, handling missing data, and using the right methods and attributes can help you work efficiently with DataFrames and avoid common errors.

Feel free to ask if you need further details or clarification on any of these points!